runtime/src/storage/iouring.rs

```diff
@@ -0,0 +1,406 @@
+        name: &[u8],
+        file: File,
+        io_sender: mpsc::Sender<(SqueueEntry, oneshot::Sender<i32>)>,
+    ) -> Self {
+        Self {
+            partition,
+            name: name.to_vec(),
+            file: Arc::new(file),
+            io_sender,
+        }
+    }
+}
+
+impl crate::Blob for Blob {
+    async fn read_at(&self, buf: &mut [u8], offset: u64) -> Result<(), Error> {
+        let fd = types::Fd(self.file.as_raw_fd());
+        let mut total_read = 0;
+
+        let mut io_sender = self.io_sender.clone();
+        while total_read < buf.len() {
+            // Figure out how much is left to read and where to read into
+            let remaining = &mut buf[total_read..];
+            let offset = offset + total_read as u64;
+
+            // Create an operation to do the read
+            let op = opcode::Read::new(fd, remaining.as_mut_ptr(), remaining.len() as _)
+                .offset(offset as _)
+                .build();
+
+            // Submit the operation
+            let (sender, receiver) = oneshot::channel();
+            io_sender
+                .send((op, sender))
+                .await
+                .map_err(|_| Error::ReadFailed)?;
+
+            // Wait for the result
+            let bytes_read = receiver.await.map_err(|_| Error::ReadFailed)?;
+
+            // If the return value is non-positive, it indicates an error.
+            let bytes_read: usize = bytes_read.try_into().map_err(|_| Error::ReadFailed)?;
+            if bytes_read == 0 {
+                // Got EOF before filling buffer.
+                return Err(Error::BlobInsufficientLength);
+            }
+            total_read += bytes_read;
+        }
+
+        Ok(())
+    }
+
+    async fn write_at(&self, buf: &[u8], offset: u64) -> Result<(), Error> {
+        let fd = types::Fd(self.file.as_raw_fd());
+        let mut total_written = 0;
+
+        let mut io_sender = self.io_sender.clone();
+        while total_written < buf.len() {
+            // Figure out how much is left to write and where to write from
+            let remaining = &buf[total_written..];
+            let offset = offset + total_written as u64;
+
+            // Create an operation to do the write
+            let op = opcode::Write::new(fd, remaining.as_ptr(), remaining.len() as _)
+                .offset(offset as _)
+                .build();
+
+            // Submit the operation
+            let (sender, receiver) = oneshot::channel();
+            io_sender
+                .send((op, sender))
+                .await
+                .map_err(|_| Error::WriteFailed)?;
+
+            // Wait for the result
+            let return_value = receiver.await.map_err(|_| Error::WriteFailed)?;
+
+            // If the return value is non-positive, it indicates an error.
+            let bytes_written: usize = return_value.try_into().map_err(|_| Error::WriteFailed)?;
+            if bytes_written == 0 {
+                return Err(Error::WriteFailed);
```

Comment:
Is it possible 0 bytes could be written for some reason (and that we should just `continue` instead of erroring)?

Reply:
`opcode::Write`'s spec says:

> Issue the equivalent of a pread(2) or pwrite(2) system call

The [man page](https://man7.org/linux/man-pages/man2/pread.2.html) for `pread` and `pwrite` says:

> On success, pread() returns the number of bytes read (a return of zero indicates end of file) and pwrite() returns the number of bytes written.

For the read case, we should never hit an EOF before reading the entire `buf`. That code should stay as is.

For the write case, from what I read it is very _unlikely_ that you will get a return value of 0, but it is technically allowed by the spec. So I will change that to only error on negative return value.